### PR TITLE
Added snapshots

### PR DESCRIPTION
**Pull request by LizaTretyakova**
The snapshot is taken from the source VM before cloning. This required me to introduce one more step for setup. I understand that it's not a very meaningful step for a user but otherwise I would have to create some elements of common environment, which are needed in other steps irrespective to whether we want to create a snapshot or not, inside the snapshot step. It would be strange.

### CODE REVIEW

File: step_clone_vm.go [hunk 1]
```diff
@@ -97,6 +80,7 @@ func (s *StepCloneVM) Run(state multistep.StateBag) multistep.StepAction {
 	}
 
 	state.Put("vm", vm)
+	state.Put("vmSrc", vmSrc)
```
Comment (mkuzmin) [inline step_clone_vm.go]: should we overwrite the previous value?

File: builder.go [hunk 2]
```diff
@@ -35,9 +35,16 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 
 	// Build the steps.
 	steps := []multistep.Step{
+		&StepSetup{
```
Comment (mkuzmin) [inline builder.go]: 👍

File: step_setup.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+	vmSrc, err := finder.VirtualMachine(ctx, s.config.Template)
+	if err != nil {
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	state.Put("client", client)
+	state.Put("ctx", ctx)
+	state.Put("finder", finder)
+	state.Put("dc", dc)
```
Comment (mkuzmin) [inline step_setup.go]: do not need it anymore - a part of finder object

File: step_create_snapshot.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import (
+	"github.com/mitchellh/multistep"
+	"github.com/hashicorp/packer/packer"
+	"context"
+	"github.com/vmware/govmomi/object"
+)
+
+type StepCreateSnapshot struct{
+	createSnapshot bool
+	vmName         string
```
Comment (mkuzmin) [inline step_create_snapshot.go]: do not need it

File: step_create_snapshot.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+func (s *StepCreateSnapshot) Run(state multistep.StateBag) multistep.StepAction {
+	ui := state.Get("ui").(packer.Ui)
+	vmSrc := state.Get("vmSrc").(*object.VirtualMachine)
+	ctx := state.Get("ctx").(context.Context)
+
+	if s.createSnapshot {
+		ui.Say("creating snapshot...")
+
+		_, err := vmSrc.CreateSnapshot(ctx, s.vmName, "", true, true)
```
Comment (mkuzmin) [inline step_create_snapshot.go]: the second parameter is not a vm name, that's a snapshot name. we can hardcode it with something like `Created by Packer` string

File: builder.go [hunk 6]
```diff
@@ -35,9 +35,16 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
...
+			config: b.config,
+		},
 		&StepCloneVM{
 			config: b.config,
 		},
+		&StepCreateSnapshot{
```
Comment (mkuzmin) [inline builder.go]: the goal of the parameter is making output VM suitable for linked clones.
If I set `cpus=2` and `create_snapshot=true` then the hardware customization is lost: the template has the setting, but child VMs do not inherit it.
Need to move snapshotting to a last action before final converting to a template.

File: step_setup.go (hunk 3 above)
Comment (LizaTretyakova) [inline step_setup.go]: I didn't find the proper getter in the `finder` so left it as-is.

**Pull request merged**